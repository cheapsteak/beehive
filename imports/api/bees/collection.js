// Framework
import { Meteor } from 'meteor/meteor';
import { Mongo } from 'meteor/mongo';
import SimpleSchema from 'simpl-schema';

class BeesCollection extends Mongo.Collection {
  insert(bee, callback) {
    // How many bee documents exist?
    const beesCursor = Bees.find({});
    const numberOfBees = beesCursor.count();
    const newNumberOfBees = numberOfBees + 1;
    // What should the name of the new bee be?
    const beeName = bee.name ? bee.name : `Bee #${newNumberOfBees}`;
    // What will the final state of the new bee object be?
    const alteredBee = { ...bee, name: beeName };
    // What is the id of the new bee object?
    const result = super.insert(alteredBee, callback);
    return result;
  }
}

export const Bees = new BeesCollection('Bees');

export default Bees;
