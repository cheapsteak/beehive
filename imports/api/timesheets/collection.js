// Framework
import { Meteor } from "meteor/meteor";
import { Mongo } from "meteor/mongo";
import SimpleSchema from "simpl-schema";

class TimesheetsCollection extends Mongo.Collection {
  insert(timesheet, callback) {
    // What will the final state of the new timesheet object be?
    const alteredTimesheet = { ...timesheet, uploaded: new Date() };
    // What is the id of the new timesheet object?
    const result = super.insert(alteredTimesheet, callback);
    return result;
  }
}

export const Timesheets = new TimesheetsCollection("Timesheets");

export default Timesheets;
