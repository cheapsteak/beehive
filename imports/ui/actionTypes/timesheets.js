export const FETCH_TIMESHEETS_REQUEST = "TIMESHEETS/FETCH_TIMESHEETS_REQUEST";
export const TIMESHEETS_RECEIVED = "TIMESHEETS/TIMESHEETS_RECEIVED";
export const TIMESHEETS_REQUEST_FAILED = "TIMESHEETS/TIMESHEETS_REQUEST_FAILED";

export const CREATE_TIMESHEET_REQUEST = "TIMESHEETS/CREATE_TIMESHEET_REQUEST";
export const CREATE_TIMESHEET_SUCCESSFUL =
  "TIMESHEETS/CREATE_TIMESHEET_SUCCESSFUL";
export const CREATE_TIMESHEET_FAILED = "TIMESHEETS/CREATE_TIMESHEET_FAILED";

export const DISMISS_TIMESHEET_ERROR = "TIMESHEETS/DISMISS_TIMESHEET_ERROR";
