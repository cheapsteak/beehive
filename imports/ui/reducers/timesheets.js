// Action Types
import {
  FETCH_TIMESHEETS_REQUEST,
  TIMESHEETS_RECEIVED,
  TIMESHEETS_REQUEST_FAILED,
  CREATE_TIMESHEET_REQUEST,
  CREATE_TIMESHEET_SUCCESSFUL,
  CREATE_TIMESHEET_FAILED,
  DISMISS_TIMESHEET_ERROR
} from "../actionTypes/timesheets";

// Default State
export const DEFAULT_STATE = {
  timesheets: [],
  error: null
};

// Reducer
export const TimesheetsReducer = (state = DEFAULT_STATE, action) => {
  switch (action.type) {
    case TIMESHEETS_RECEIVED:
      return {
        ...state,
        timesheets: action.payload
      };
    case CREATE_TIMESHEET_FAILED:
    case TIMESHEETS_REQUEST_FAILED:
      return {
        ...state,
        error: action.payload
      };
    case DISMISS_TIMESHEET_ERROR:
      return {
        ...state,
        error: null
      };
    case CREATE_TIMESHEET_SUCCESSFUL:
    case FETCH_TIMESHEETS_REQUEST:
    case CREATE_TIMESHEET_REQUEST:
    default:
      return state;
  }
};

export default TimesheetsReducer;
