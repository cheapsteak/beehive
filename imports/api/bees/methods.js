// Framework
import { Meteor } from 'meteor/meteor';
import SimpleSchema from 'simpl-schema';
import { ValidatedMethod } from 'meteor/mdg:validated-method';
import { DDPRateLimiter } from 'meteor/ddp-rate-limiter';

// Libraries
import _ from 'underscore';

// Collections
import { Bees } from './collection';

export const fetchAllForUser = new ValidatedMethod({
  name: 'bees.fetchAllForUser',
  validate: null,
  run() {
    const allBeesForUserCursor = Bees.find({ userId: this.userId });
    const allBeesForUser = allBeesForUserCursor.fetch();
    return allBeesForUser;
  }
});

export const insert = new ValidatedMethod({
  name: 'bees.insert',
  validate: null,
  run({ bee }) {
    try {
      const newBee = {
        ...bee,
        userId: this.userId
      };
      const result = Bees.insert(newBee);
      return result;
    } catch (error) {
      const newError = new Meteor.Error(
        'api.bees.insert.unspecifiedError',
        'Could not insert a new bee.',
        error
      );
      console.error(newError);
      throw newError;
    }
  }
});

export const updateName = new ValidatedMethod({
  name: 'bees.updateName',
  validate: new SimpleSchema({
    beeId: { type: String },
    newName: { type: String }
  }).validator(),
  run({ beeId, newName }) {
    const bee = Bees.findOne(beeId);

    //TODO: fix hack, unsuccessful add editableBy function through collection.js
    if (false && !bee.editableBy(this.userId)) {
      throw new Meteor.Error(
        'api.bees.updateName.accessDenied',
        "You don't have permission to edit this bee."
      );
    }

    Bees.update(beeId, {
      $set: { name: newName }
    });
  }
});

export const remove = new ValidatedMethod({
  name: 'bees.remove',
  validate: new SimpleSchema({
    beeId: { type: String }
  }).validator(),
  run({ beeId }) {
    const bee = Bees.findOne(beeId);

    //TODO: fix hack, unsuccessful add editableBy function through collection.js
    if (false && !bee.editableBy(this.userId)) {
      throw new Meteor.Error(
        'api.bees.remove.accessDenied',
        "You don't have permission to remove this bee."
      );
    }

    return Bees.remove(beeId);
  }
});

export const removeAllForUser = new ValidatedMethod({
  name: 'bees.removeAllForUser',
  validate: null,
  run() {
    return Bees.remove({ userId: this.userId });
  }
});

const BEES_METHODS = _.pluck(
  [insert, updateName, remove, removeAllForUser, fetchAllForUser],
  'name'
);

if (Meteor.isServer) {
  DDPRateLimiter.addRule(
    {
      name: name => _.contains(BEES_METHODS, name),
      connectionId: () => true
    },
    50, // Operations per limiting interval
    1000 // Limiting interval
  );
}
