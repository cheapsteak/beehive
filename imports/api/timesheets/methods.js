// Framework
import { Meteor } from "meteor/meteor";
import { ValidatedMethod } from "meteor/mdg:validated-method";
import { DDPRateLimiter } from "meteor/ddp-rate-limiter";
import SimpleSchema from "simpl-schema";
import moment from "moment";
import csv from "csv";

// Libraries
import _ from "underscore";

// Collections
import { Timesheets } from "./collection";

export const fetchAll = new ValidatedMethod({
  name: "timesheets.fetchAll",
  validate: null,
  run() {
    return Timesheets.find({}, { sort: { date: -1 } }).fetch();
  }
});

export const insert = new ValidatedMethod({
  name: "timesheets.insert",
  validate: new SimpleSchema({
    timesheet: { type: String }
  }).validator(),
  run({timesheet}) {
    const entries = Meteor.wrapAsync(csv.parse)(timesheet);
    if (entries.length < 2) {
      throw new Meteor.Error("time does not have enough rows to be parsed");
    }

    //the last entry in index spot 1 is the report id
    //popping it removes it from entries
    const headers = entries.shift();
    const reportId = entries.pop()[1];

    const existingReport = Timesheets.findOne({ reportId });
    if (existingReport) {
      throw new Meteor.Error(`Report #${reportId} has already been uploaded.`);
    }

    return entries.map(([workDate, hours, employee, job]) =>
      Timesheets.insert({
        reportId,
        date: moment(workDate, "DD/MM/YYYY").toDate(),
        hours: parseInt(hours),
        employee,
        job
      })
    );
  }
});

const TIMESHEETS_METHODS = _.pluck([insert, fetchAll], "name");

if (Meteor.isServer) {
  DDPRateLimiter.addRule(
    {
      name: name => _.contains(TIMESHEETS_METHODS, name),
      connectionId: () => true
    },
    50, // Operations per limiting interval
    1000 // Limiting interval
  );
}
