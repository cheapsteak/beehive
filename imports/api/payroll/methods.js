// Framework
import { Meteor } from "meteor/meteor";
import { ValidatedMethod } from "meteor/mdg:validated-method";
import { DDPRateLimiter } from "meteor/ddp-rate-limiter";
import _ from "underscore";
import moment from "moment";
import csv from "csv";

// Collections
import { Timesheets } from "../timesheets/collection";

const jobPrices = {
  A: 20,
  B: 30
};

export const fetchAll = new ValidatedMethod({
  name: "payroll.fetchAll",
  validate: null,
  run() {
    const timesheets = Timesheets.find(
      {},
      { sort: { date: -1, employee: -1 } }
    ).fetch();

    //TODO: this is effective, but messy
    return (
      _.chain(timesheets)
        .groupBy(({ date }) => {
          // determines which entry is in which pay period
          const timestamp = moment(date);
          const monthYear = timestamp.format("/MM/YYYY");
          if (timestamp.date() < 16) {
            return `1${monthYear}-15${monthYear}`;
          }
          return `16${monthYear}-${timestamp.daysInMonth()}${monthYear}`;
        })
        //group by employee, total up payments
        .mapObject((timeEntries, period) =>
          _.chain(timeEntries)
            .groupBy("employee")
            .mapObject((employeeTimeEntries, employee) => ({
              period: period
                .split("-")
                .map(date => moment(date, "DD/MM/YYYY").toDate()),
              employee,
              amount: employeeTimeEntries.reduce(
                (total, { hours, job }) => jobPrices[job] * hours + total,
                0
              )
            }))
            .values()
            .value()
        )
        .values()
        .flatten()
        .value()
    );
  }
});

const PAYROLL_METHODS = _.pluck([fetchAll], "name");

if (Meteor.isServer) {
  DDPRateLimiter.addRule(
    {
      name: name => _.contains(PAYROLL_METHODS, name),
      connectionId: () => true
    },
    50, // Operations per limiting interval
    1000 // Limiting interval
  );
}
