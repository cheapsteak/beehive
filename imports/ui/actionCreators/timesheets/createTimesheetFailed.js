// Actoon Types
import { CREATE_TIMESHEET_FAILED } from "../../actionTypes/timesheets";

// Action Creator
export const createTimesheetFailed = error => ({
  type: CREATE_TIMESHEET_FAILED,
  payload: error
});

export default createTimesheetFailed;
