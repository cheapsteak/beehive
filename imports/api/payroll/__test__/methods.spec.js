import { Meteor } from "meteor/meteor";
import { Mongo } from "meteor/mongo";
import { ValidatedMethod } from "meteor/mdg:validated-method";

import moment from "moment";

import { Timesheets } from "../../timesheets/collection";

import { fetchAll } from "../methods";

describe("payroll methods", () => {
  let collection;
  beforeEach(() => {
    collection = Mongo.Collection.prototype;
  });

  describe("payroll.fetchAll", () => {
    it("runs normally", () => {
      const makeMoment = dateString =>
        moment(dateString, "DD/MM/YYYY").toDate();
      collection.find = jest.fn(props => {
        return {
          fetch: () => [
            {
              date: makeMoment("4/11/2016"),
              employee: "1",
              hours: 10,
              job: "A"
            },
            {
              date: makeMoment("14/11/2016"),
              employee: "2",
              hours: 4,
              job: "B"
            },
            {
              date: makeMoment("20/11/2016"),
              employee: "1",
              hours: 3,
              job: "B"
            }
          ]
        };
      });

      const result = fetchAll.run();
      expect(Timesheets.find).toHaveBeenCalledWith(
        {},
        { sort: { date: -1, employee: -1 } }
      );

      expect(result).toMatchSnapshot();

      expect(result[0].amount).toBe(200);
      expect(result[0].employee).toBe("1");

      expect(result[1].amount).toBe(120);
      expect(result[1].employee).toBe("2");
    });
  });
});
