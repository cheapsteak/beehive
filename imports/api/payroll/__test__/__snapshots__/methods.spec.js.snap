// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payroll methods payroll.fetchAll runs normally 1`] = `
Array [
  Object {
    "amount": 200,
    "employee": "1",
    "period": Array [
      2016-11-01T04:00:00.000Z,
      2016-11-15T05:00:00.000Z,
    ],
  },
  Object {
    "amount": 120,
    "employee": "2",
    "period": Array [
      2016-11-01T04:00:00.000Z,
      2016-11-15T05:00:00.000Z,
    ],
  },
  Object {
    "amount": 90,
    "employee": "1",
    "period": Array [
      2016-11-16T05:00:00.000Z,
      2016-11-30T05:00:00.000Z,
    ],
  },
]
`;
